app/services/order_invoice_comparator.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+class OrderInvoiceComparator
+  def can_generate_new_invoice?(current_state_invoice, latest_invoice)
+    # We'll use a recursive BFS algorithm to find if the invoice is outdated
+    # the root will be the order
+    # On each node, we'll a list of relevant attributes that will be used on the comparison
+    bfs(current_state_invoice.presenter, latest_invoice.presenter, invoice_generation_selector)
+  end
+
+  def can_update_latest_invoice?(current_state_invoice, latest_invoice)
+    bfs(current_state_invoice.presenter, latest_invoice.presenter, invoice_update_selector)
+  end
+
+  def bfs(node1, node2, attributes_selector)
```

Comment:
`bfs` is the type of algorithm but it would be nice if the name told us what the return value will be. I would suggest:
```suggestion
  def different?(node1, node2, attributes_selector)
```